borg/archiver.py

```diff
@@ -353,6 +355,155 @@ def do_extract(self, args):
+                        print("\t", args.compare, compare_item.get(b'source', '<regular file>'))
+                return
+            if item[b'chunks'] != compare_item[b'chunks'] or deleted or not can_compare_chunks:
+                # Contents are different
+                ref_chunks = {c[0] for c in item[b'chunks']}
+                cmp_chunks = {c[0] for c in compare_item[b'chunks']}
+                total_ref = None if not item[b'mode'] else sum(c[1] for c in item[b'chunks'])
+                total_compare = None if not compare_item[b'mode'] else sum(c[1] for c in compare_item[b'chunks'])
+                if (not can_compare_chunks and total_ref == total_compare and not deleted and
+                        compare_chunks(ref_chunks, cmp_chunks, ref, compare)):
+                    return
+                added = sum(c[1] for c in (cmp_chunks - ref_chunks))
+                removed = sum(c[1] for c in (ref_chunks - cmp_chunks))
+                if args.json:
+                    dump_json({
+                        "path": path,
+                        "what": "content",
+                        "added": added,
+                        "removed": removed,
+                        "total_reference": total_ref,
+                        "total_compare": total_compare,
+                    })
+                else:
+                    print(remove_surrogates(path), "different contents")
+                    print("\t+%s, -%s, ref %s, compare %s" % (format_bytes(added), format_bytes(removed),
+                                                              format_bytes(total_ref), format_bytes(total_compare)))
+
+        repository = self.open_repository(args)
+        manifest, key = Manifest.load(repository)
+        ref = Archive(repository, key, manifest, args.location.archive)
+        compare = Archive(repository, key, manifest, args.compare)
+
+        # Extract and check chunker params
+        cmdline_parser = self.build_parser()
+        ref_chunker_params = ref.metadata.get(b'chunker_params', cmdline_parser.parse_args(
+            ref.metadata[b'cmdline'][1:]).chunker_params)
+        cmp_chunker_params = compare.metadata.get(b'chunker_params', cmdline_parser.parse_args(
+            compare.metadata[b'cmdline'][1:]).chunker_params)
+        can_compare_chunks = ref_chunker_params == cmp_chunker_params
+        if not can_compare_chunks:
+            self.print_warning("--chunker-params are different between archives, diff will be slow")
+
+        matcher = PatternMatcher()
+        if args.excludes:
+            matcher.add(args.excludes, False)
+        include_patterns = []
+        if args.paths:
+            include_patterns.extend(parse_pattern(i, PathPrefixPattern) for i in args.paths)
+            matcher.add(include_patterns, True)
+        matcher.fallback = not include_patterns
+
+        if args.json:
+            print("[")
+        orphan_refs = {}
+        orphan_cmps = {}
+        for refitem, cmpitem in zip_longest(
```

Comment:
Yes, but it's still very fast, because the orphan_* dicts only hold few items (in the test case I am using the peak size is about 6000, of which 5500 persist after this loop, so are never consumed in this loop, because these 5500 where deleted in either archive).

Using a list might be faster in some cases ("cold" end with deleted orphans, "hot" end which is only as large as the difference in the file lists). I take a look at that.